test/upgrade/fleet_test.go

```diff
@@ -0,0 +1,373 @@
+package upgrade
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/fleetdm/fleet/v4/server/service"
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/jmoiron/sqlx"
+)
+
+// Slots correspond to docker-compose fleet services, either fleet-a or fleet-b
+const (
+	slotA = "a"
+	slotB = "b"
+)
+
+func init() {
+	rand.Seed(time.Now().Unix())
+}
+
+// Fleet represents the fleet server and its dependencies used for testing.
+type Fleet struct {
+	ProjectName string
+	FilePath    string
+	Version     string
+	ActiveSlot  string
```

Comment:
Seems like this field is not used

Reply:
Removed. I tried to make it possible to upgrade multiple times, and have `ActiveSlot` alternate between `a` and `b`, but it was turning out to be a little more complex than I hoped. If we need to support multiple upgrades, we can revisit this.